### ISSUE TITLE
Feature: support read_csv_auto reading from the '/dev/stdin' file

### ISSUE THREAD

**Issue by mskyttner** [duckdb/duckdb#1417]
The duckdb CLI command can read from a file using read_csv_auto():

```bash
duckdb mydb "create table mytable as select * from read_csv_auto('mydata.csv', HEADER=TRUE);"
```

When I try to send the same data to the /dev/stdin file and have duckdb load it from there, it fails:

```bash
cat mydata.csv | duckdb mydb "create table mytable as select * from read_csv_auto('/dev/stdin', HEADER=TRUE);" 
```

The error is:

```bash
Error: Invalid Input Error: Error in file "/dev/stdin": CSV options could not be auto-detected. Consider setting parser options manually.
```

**Comment by hannesmuehleisen** [comment]
Maybe @tdoehmen can  have a look?

**Comment by mskyttner** [comment]
Just sharing some thoughts on the use cases. The duckdb CLI shell offers similar capabilities as sqlite3 for moving data in and out using both special "dot commands" and "SQL" command statements. Automating loading and export from the commandline, some of these so called dot-commands are supported: https://sqlite.org/cli.html... 

I think some common dot commands are `.import`, another is `.dump`, for reading and writing data in and out of the process. There is also `.output` (with different `.mode` settings). It seems those dot commands in sqlite3 have grown organically and there are now quite a few of these in sqlite3. In duckdb CLI shell, there seems to be support for `.import` and other dot commands but not a 1-to-1 match with all sqlite3 dot commands.

It seems there are faster and better duckdb-"sql"-statements for reading and writing data into and out from the duckdb process, such as maybe using:

- "create table if not exists stdin as select * from read_csv_auto('/dev/stdin')" # for reading data 
- "export database '/dev/std' (FORMAT PARQUET);" # for writing data output (binary)
- "copy stdin to '/dev/stdout' (FORMAT CSV);" # for writing data in a table called stdin to an output file (in text format)

When reading and writing streams of data coming from stdin and going to stdout, it would be nice to have easy access to these files/streams, to use some of these faster duckdb specific import and export functions with less intermediate steps. It would be nice to not have to create various temp files and directories, unless required, for example due to larger data which cannot fit in RAM.

The dot commands can be piped to sqlite3, but not easily "data" so duckdb would have an edge if load and dump functions were to work with data from stdin, and results going to stdout. For sqlite3, to alleviate this, there are tools like https://tobimensch.github.io/termsql/ provided outside of sqlite3. Or shell wrappers can be used; here are some links that describe some usage/workarounds with sqlite3:
  
- https://unix.stackexchange.com/questions/288273/bash-redirection-doesnt-work-for-sqlite-command
- https://stackoverflow.com/questions/41354486/sqlite-how-to-simultaneously-read-data-from-stdin-and-table-name-from-variable

Here is a [rough attempt to do an R script](https://gist.github.com/mskyttner/46601e60138a886cdfbfc81067c71725) "outside of duckdb shell/CLI" which reads from stdin and writes to stdout. It is quite quick and dirty, just adding it to illustrate use cases that could be nice to support on the commandline, in my opinion, for example (after copying the script to ~/bin and doing chmod +x) ...

```bash
df | tail -n +2 | duckstream.R --no-header -d " " -Q "select X1,X2 from stdin;" --dump myparquetoutput
```

**Comment by tdoehmen** [comment]
@hannesmuehleisen I am not familiar with the cli integration of duckdb, so I only have a limited view on this. 
In the csv reader, I could imagine that it's possible to treat `path=/dev/stdin` as magic value, which makes the parser read from stdin with something like `std::getline(std::cin,line)`. The parser would need to treat it similar to a .gz file (not the compression, but that data has to be read sequentially and that we can't jump to other positions in the input), which means that dialect and data type detection for CSV will only work to a limited extend.

**Comment by mskyttner** [comment]
Bulk loading functionality through command line tools with input from stdin is supported in some databases, for example https://mariadb.com/kb/en/columnstore-bulk-data-loading/#bulking-loading-data-from-stdin. It would be pretty nice if that was built into the duckdb shell / CLI command instead of being provided "on the outside" by a separate utility. I guess that is what some of the dot commands in sqlite do, extend the shell with input/output features.

The use cases I had in mind when opening this feature wish was not only specifically about converting from CSV to PARQUET but also about being able to avoid reading data from or to disk altogether, like in the examples given here: https://github.com/tobimensch/termsql/#examples. But specifically for the example use case with exporting to parquet ... I admit that I use the CSV->duckdb->PARQUET a lot now, until the duckdb file format settles. 

Would it be possible to buffer the data from stdin in chunks into a "buffer" duckdb table (maybe single column, varchar), with chunkwise parse/move operations to final destination happening from there? First chunk would be a lookahead of n lines, used to guess data types, separators etc (if not specified), and all subsequent chunks of m lines each would be parsed with those settings?

If someone sends an endless stream (by mistake or as a Denial-of-Service attempt), it would be great to see some progress and finally warning / abort before system resources get depleted.

**Comment by tdoehmen** [comment]
Thank you for the clarification and the information. Then, a more general solution would indeed make more sense.  

Maybe buffering to a table could indeed save some headache around disk-spilling! It doesn't feel like a super clean solution (i'd prefer to buffer the raw data - if at all), but maybe it's a good starting point. One option to go without buffering would be to allow the use of stdin input only for specific queries which are guaranteed to need the input only once (e.g. COPY ... INTO). 

Endless input streams would, however, still remain a problem, thanks for the hint. I think that a line-break could generally be used to mark the end of the stream. Maybe an option like max_lines could be added to prevent endless reading from the stdin-stream if the line-break never comes. 

I can unfortunately not promise an immediate solution to this, as I am not familiar with the CLI/shell integration yet and also have to give the buffering aspect another thought.  But I will investigate and come back to you when I hopefully know more ; )

**Comment by dforsber** [comment]
I would highly value (low level) streaming in and out APIs for DuckDB. First, streaming in general speeds up data processing and would enable nice stream processing use cases, buffer as much as needed/wanted, e.g. for window functions, but generally streaming allows processing huge amounts of data with small amounts of memory.

Furthermore, as a generalisation, DuckDB could read streams as any tables, and cache them into (compressed) memory if full "scan" is needed for column(s). CSV is anyway a (linear?) data source as well (varying row lengths). It would be badass cool to chain DuckDB to DuckDB.

**Comment by hannesmuehleisen** [comment]
We have streaming APIs both for reading table data and to retrieve results. Could you be more precise what you are missing?

**Comment by dforsber** [comment]
> We have streaming APIs both for reading table data and to retrieve results. Could you be more precise what you are missing?

My knowledge of the APIs is still limited, and more around node module we have. I'm thinking about query pushdown from DuckDB to DuckDB and being able to stream the response in a "standard" data format that is easily consumable (e.g. the apache arrow flight that was referred to in some other issue). Almost like read replication, but instead of full data replication, just stream based on a query and having the received data in a queryable format for in-memory DuckDB table for example.

**Comment by pkit** [comment]
> We have streaming APIs both for reading table data and to retrieve results. Could you be more precise what you are missing?

@hannesmuehleisen

I can explain what's missing.
DuckDB uses `seek` to check for csv type and auto detect stuff.
But that wouldn't work with streamable csv files, for example named pipes or http sources.
For the use case of: "load csv into RAM from external source and then do some queries" it becomes cumbersome pretty fast.
It expects fully functional POSIX file system with random access to files __just__ to load the file into RAM (`CanSeek()` always returns `true`, hardcoded), which is a pretty far reaching assumption.
I can probably do a PR on how to fix that, but I'm not sure it's a wanted change, judging by no support for STDIN sources.

Contrived example of what I'm doing:
```python
import os
import tempfile
import duckdb
from multiprocessing import Process

tmpdir = tempfile.mkdtemp()
data_file = os.path.join(tmpdir, 'data.csv')
os.mkfifo(data_file)

con = duckdb.connect(database=':memory:', read_only=False)

con.execute("DROP TABLE if exists cities")
con.execute("CREATE TABLE cities (latd integer, latm integer, lats integer, ns varchar, lond integer, lonm integer, lons integer, ew varchar, city varchar, state varchar)")


def pipe_data(src, dst):
    fifo = open(dst, 'wb')
    with open(src, 'rb') as file:
        while chunk := file.read(1024):
            fifo.write(chunk)


def load_data(event, context):
    path = 'cities.csv'
    process = Process(target=pipe_data, args=(path, data_file,))
    process.start()

    con.execute('COPY cities FROM \'{file_name}\' ( HEADER )'.format(file_name=data_file))
    process.join()
```

I'm open to hear what I can do better here.

**Comment by Mytherin** [comment]
> I can explain what's missing.
> DuckDB uses `seek` to check for csv type and auto detect stuff.
> But that wouldn't work with streamable csv files, for example named pipes or http sources.
> For the use case of: "load csv into RAM from external source and then do some queries" it becomes cumbersome pretty fast.
> It expects fully functional POSIX file system with random access to files **just** to load the file into RAM (`CanSeek()` always returns `true`, hardcoded), which is a pretty far reaching assumption.

If `CanSeek()` returns true, the CSV reader uses seek to jump around in the file to look for samples for better auto detection. If `CanSeek()` is false it will instead only read from the beginning of the stream to do the type detection.

It is true that the regular file system (which is supposed to be a"standard/POSIX file system) always returns true for CanSeek. This is not hard-coded, however. The method is virtual and can be overriden/replaced. For example, the [gzip file system](https://github.com/duckdb/duckdb/blob/30fda3efeb7ca07390a84e787c3110ede289153f/src/include/duckdb/common/gzip_file_system.hpp) returns false here (since random seeking inside gzip files is difficult/expensive). 

For pipes I would recommend using a new file system as well, since pipes are (as you mentioned) not a regular file system and do not support the same set of operations that regular file systems do.

I do still see one problem in supporting pipes there, however. The file system will call the `Reset` method on the file, which will still be problematic for pipes. Instead, ideally the CSV reader would need a minor redesign so it would re-use its own internal buffer instead of calling reset.

Also, minor note, CSV files [already work with HTTP sources](https://github.com/duckdb/duckdb/blob/30fda3efeb7ca07390a84e787c3110ede289153f/test/sql/copy/csv/test_csv_remote.test) ;)

> I can probably do a PR on how to fix that, but I'm not sure it's a wanted change, judging by no support for STDIN sources.

Supporting pipes for CSV file reading (and in general) seems like a good idea, no objections from my side. Happy to review a PR on this!

**Comment by pkit** [comment]
@Mytherin 
Working http sources is pretty nice.
But for `PipeFileSystem` I'm not sure how it should be chosen? In posix a pipe is just a file. It has no specific schema like http or extension like gzip.
Or better, can you point me in the direction of the code that does the "fle system specialization"?

**Comment by Mytherin** [comment]
> Or better, can you point me in the direction of the code that does the "fle system specialization"?

The file system specialization happens [here](https://github.com/duckdb/duckdb/blob/14193fb87eb88566c44a417a73c3e50a4b0da8cc/src/common/file_system.cpp#L1014) for gzip files.

> P.S. and the Reset() comment is spot on. The buffered csv reader cannot reset stream at will, but it's quite happily doing it right now :)

That is true and should probably be changed. However, do note that the auto-detect is optional and can be disabled. So even without any changes to the code streaming through a pipe might work already if auto-detect is disabled and the schema is manually specified.

**Comment by pkit** [comment]
> The file system specialization happens here for gzip files.

Hmm, sorry to bother you so much, but where it decides if `http` should be used then?

> That is true and should probably be changed. However, do note that the auto-detect is optional and can be disabled. So even without any changes to the code streaming through a pipe might work already if auto-detect is disabled and the schema is manually specified.

Tried defining the schema and it still tried to seek it.
I think auto_detect is off by default?
It looks like it will `Reset()` anyway here: https://github.com/duckdb/duckdb/blob/45c9cb51852ecd14338f79854fe898d70054e8e7/src/execution/operator/persistent/buffered_csv_reader.cpp#L140

**Comment by Mytherin** [comment]
> Hmm, sorry to bother you so much, but where it decides if http should be used then?

That happens in the `FindFileSystem(path)`, which uses the `CanHandleFile` to figure out which file system to use ([which is overloaded here for the httpfs](https://github.com/duckdb/duckdb/blob/30fda3efeb7ca07390a84e787c3110ede289153f/extension/httpfs/httpfs.cpp#L180)).

> I think auto_detect is off by default?

It is turned on by default, but indeed it seems that the JumpToBeginning is still called in either case. That can be changed as a reset should not be required in that scenario.

**Comment by pkit** [comment]
>It is turned on by default, but indeed it seems that the JumpToBeginning is still called in either case. That can be changed as a reset should not be required in that scenario.

Yup. 

It also looks like the http loader is not in the latest (0.2.7) release yet?

```python
http_url = 'https://people.sc.fsu.edu/~jburkardt/data/csv/cities.csv'
con.execute('COPY cities FROM \'{file_name}\' ( HEADER )'.format(file_name=http_url))
```

```bash
RuntimeError: IO Error: No files found that match the pattern "https://people.sc.fsu.edu/~jburkardt/data/csv/cities.csv"
```

**Comment by Mytherin** [comment]
It is supported, but you need to enable the httpfs extension (`BUILD_HTTPFS=1 make` if you are building from source, or add it to the extensions list in `tools/pythonpkg/setup.py`). It is not enabled by default because of the dependency on OpenSSL.

We are still working on making it easier to have extensions loaded externally without needing to compile from source, so you could do e.g. `pip install duckdb-httpfs` and install the extension separately. That is not available yet unfortunately.

**Comment by mskyttner** [comment]
Regarding the use case to "load csv into RAM from external source and then do some queries" using `read_csv_auto`, the automatic type guessing enabled by default is very useful and convenient. I can read a csv from a web url with the duckdb CLI and output results to stdout from a select query in one big chunk:

```bash
curl -s https://raw.githubusercontent.com/duckdb/duckdb/master/test/sql/copy/csv/data/real/web_page.csv | sed 's/|/,/g' > /tmp/stdin \
&& cat /tmp/stdin | duckdb -csv :memory: "create table mytable as select * from read_csv_auto('/tmp/stdin', HEADER=FALSE); select * from mytable;"
```

With duckdb CLI v0.2.7 this works so httpsfs is supported there:

```bash
duckdb -csv :memory: "create table mytable as select * from read_csv_auto('https://raw.githubusercontent.com/duckdb/duckdb/master/test/sql/copy/csv/data/real/web_page.csv', HEADER=FALSE, SEP='|'); select * from mytable;"
```

If the duckdb CLI could read directly from the stdin stream, maybe even with a progress indicator for chunkwise processing steps, there would be no need to create extra temp files or write external utilities to load data incrementally in chunks (as in https://github.com/duckdb/duckdb/issues/1417#issuecomment-867506887 and https://gist.github.com/mskyttner/46601e60138a886cdfbfc81067c71725).

Here is a duckdb CLI v 0.2.7 attempt to load data without the temp file:

```bash
curl -s https://raw.githubusercontent.com/duckdb/duckdb/master/test/sql/copy/csv/data/real/web_page.csv | sed 's/|/,/g' \
| duckdb -csv :memory: "create table mytable as select * from read_csv_auto('/dev/stdin', HEADER=TRUE); select * from mytable;"
```
The error reported is:

`Error: IO Error: Could not seek to location 0 for file "/dev/stdin": Illegal seek`

Would fixing the JumpToBeginning issue help here?

**Comment by pkit** [comment]
> Would fixing the JumpToBeginning issue help here?

Yup, looks like it. Just need to detect that the stream is unseekable and then chose correct fs implementation. After getting rid of `JumpToBeginning`

**Comment by pkit** [comment]
> It is supported, but you need to enable the httpfs extension (`BUILD_HTTPFS=1 make` if you are building from source, or add it to the extensions list in `tools/pythonpkg/setup.py`). It is not enabled by default because of the dependency on OpenSSL.

Hmm, I cannot build a python package though.
```bash
cd tools/pythonpkg
python setup.py sdist

Traceback (most recent call last):
  File "setup.py", line 99, in <module>
    (source_list, include_list, original_sources) = package_build.build_package(os.path.join(script_path, 'duckdb'), extensions)
  File "/home/user/git/duckdb/tools/pythonpkg/../../scripts/package_build.py", line 149, in build_package
    include_package(ext, ext_path, include_files, include_list, source_list)
  File "/home/user/git/duckdb/tools/pythonpkg/../../scripts/package_build.py", line 105, in include_package
    ext_pkg = __import__(pkg_name + '_config')
ModuleNotFoundError: No module named 'httpfs_config'

```

The build with `BUILD_HTTPFS=1 make` succeeds though

**Comment by pkit** [comment]
@mskyttner 

The following will never work, because it needs to sample data which is impossible when streaming
```
curl -s https://raw.githubusercontent.com/duckdb/duckdb/master/test/sql/copy/csv/data/real/web_page.csv | sed 's/|/,/g' \
| duckdb -csv :memory: "create table mytable as select * from read_csv_auto('/dev/stdin', HEADER=TRUE); select * from mytable;"
```
This one will work soon:
```
curl -s https://raw.githubusercontent.com/duckdb/duckdb/master/test/sql/copy/csv/data/real/web_page.csv | sed 's/|/,/g' \
| duckdb -csv :memory: "create table mytable as select * from read_csv('/dev/stdin', columns=STRUCT_PACK(foo := 'INTEGER', bar := 'VARCHAR', baz := 'VARCHAR', bam := 'VARCHAR'), AUTO_DETECT='false'); select * from mytable;"
```

**Comment by mskyttner** [comment]
Nice, this should fix the `Error: IO Error: Could not seek to location 0 for file "/dev/stdin": Illegal seek` mentioned earlier! 

Regarding the "impossible when streaming" statement when using `read_csv_auto`, can you help me to understand better the challenge there? The read_csv_auto type guessing/detection by default uses random seek across the whole file to sample rows and base type guessing on this sample, right? For a chunked line buffered stream of CSV data this would not work since we only have individual chunks of data to work with? Is this why you say it would never work to use read_csv_auto on a buffered text stream to do the type guessing for the CSV columns? 

But `read_csv_auto` allows for for type guessing using random sampling across the whole file to be switched off, since it supports [options SAMPLE_SIZE and ALL_VARCHAR](https://duckdb.org/docs/data/csv)? What about then (for a line buffered text stream) "downgrading" the type guessing strategy to either treat all columns as varchar (worst case) or to buffer the first thousand rows and base types on this sample rather than a random sample across all rows (if later chunks held lines that were "non-conformant" either the columns could be "demoted" to a less specific type such as varchar or replaced by NA while parsing issues could be logged (to stderr?) [in table format similar to how readr::read_csv does this](https://readr.tidyverse.org/reference/problems.html)? 

Or to add CHUNK_SIZE and use a combination of a lookahead and chunksize settings, such as in [this wrapper](https://gist.github.com/mskyttner/46601e60138a886cdfbfc81067c71725#file-duckstream-r-L37-L46)? This way column types and names (and number of columns) would not be need to be specified before reading the stream of CSV lines. Clean-up could later be done in separate steps, when data has landed in the database.

**Comment by pkit** [comment]
> But read_csv_auto allows for for type guessing using random sampling across the whole file to be switched off, since it supports options SAMPLE_SIZE and ALL_VARCHAR? What about then (for a line buffered text stream) "downgrading" the type guessing strategy to either treat all columns as varchar (worst case) or to buffer the first thousand rows and base types on this sample rather than a random sample across all rows (if later chunks held lines that were "non-conformant" either the columns could be "demoted" to a less specific type such as varchar or replaced by NA while parsing issues could be logged (to stderr?) in table format similar to how readr::read_csv does this?

> Or to add CHUNK_SIZE and use a combination of a lookahead and chunksize settings, such as in this wrapper? This way column types and names (and number of columns) would not be need to be specified before reading the stream of CSV lines. Clean-up could later be done in separate steps, when data has landed in the database.

Yep, both of these could be done, but it's much more involved and will need another feature (or two).
"Impossible" in a strict streaming sense where you want to keep as low of a memory footprint as possible (I usually prefer <=64KB) and just pump data through as fast as possible.

**Comment by mskyttner** [comment]
With version 0.2.8, this duckdb CLI command works (which is great!):

`curl -s https://raw.githubusercontent.com/duckdb/duckdb/master/test/sql/copy/csv/data/real/web_page.csv | sed 's/|/,/g' | duckdb -csv :memory: "create table mytable as select * from read_csv('/dev/stdin', columns=STRUCT_PACK(foo := 'INTEGER', bar := 'VARCHAR', baz := 'VARCHAR', bam := 'VARCHAR', a := 'VARCHAR', b := 'VARCHAR', c := 'VARCHAR', d := 'VARCHAR', e := 'VARCHAR', f := 'VARCHAR', g := 'VARCHAR', h := 'VARCHAR', i := 'VARCHAR', j := 'VARCHAR'), AUTO_DETECT='false'); select * from mytable;"
`

The original issue was about using `read_csv_auto` rather than `read_csv`, since it would be quite nice to avoid specifying the column types explicitly. When swapping for this in the command above it results in `Error: Not implemented Error: Unsupported: Reset pipe/stream` which is a little unexpected perhaps, since it is an alias to `read_csv` according to the documentation? 

Would it be a good idea to open another issue to try to capture those features needed as mentioned by @pkit, in order to be able to do chunked line buffered reading from stdin using duckb CLI?

It is possible to do `duckdb -csv :memory: "select * from read_csv_auto('https://raw.githubusercontent.com/duckdb/duckdb/master/test/sql/copy/csv/data/real/web_page.csv')"` which is great but does not allow "pre-processing" of the content going in using a chain of various command line utilities. 

For example, it would be nice to be able to send the output (below unchanged) further to another duckdb CLI call in another step:

`duckdb -csv :memory: "select * from read_csv_auto('https://raw.githubusercontent.com/duckdb/duckdb/master/test/sql/copy/csv/data/real/web_page.csv')" | duckdb -csv :memory: "select * from read_csv_auto('/dev/stdin');"`

**Comment by corneliusroemer** [comment]
@mskyttner I've opened your idea to enable `read_csv_auto` as a feature request in #2829

**Comment by hossam-houssien** [comment]
this not working on windows using /dev/stdin using v0.8.1 6536a77232

**Comment by hossam-houssien** [comment]
so what about Windows platform?

**Comment by derekmahar** [comment]
One solution to support standard input on all platforms might be to copy the approach that [csvq](https://mithrandie.github.io/csvq/) uses which is to define a special internal file name [`STDIN` ](https://mithrandie.github.io/csvq/reference/select-query.html#from_clause) to represent standard input.
